controller/resource_test.go

```diff
@@ -14,94 +15,286 @@ import (
+			aclTokens: []*api.ACLToken{aclToken1},
+			expected: map[string]ServiceState{
+				"service1": {
+					ConsulECSTasks: false,
+					ACLTokens:      []*api.ACLTokenListEntry{aclTokenListEntry1},
+				},
+			},
+		},
+	}
+
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			adminToken := "123e4567-e89b-12d3-a456-426614174000"
+			testServer, err := testutil.NewTestServerConfigT(t, func(c *testutil.TestServerConfig) {
+				c.ACL.Enabled = true
+				c.ACL.Tokens.Master = adminToken
+				c.ACL.DefaultPolicy = "deny"
+			})
+			require.NoError(t, err)
+			t.Cleanup(func() { _ = testServer.Stop() })
+			testServer.WaitForLeader(t)
+
+			clientConfig := api.DefaultConfig()
+			clientConfig.Address = testServer.HTTPAddr
+			clientConfig.Token = adminToken
+
+			consulClient, err := api.NewClient(clientConfig)
+			require.NoError(t, err)
+
+			for _, aclToken := range c.aclTokens {
+				_, _, err = consulClient.ACL().TokenCreate(aclToken, nil)
+				require.NoError(t, err)
+			}
+
+			var tasks []*ecs.Task
+
+			for i := range c.tasks {
+				tasks = append(tasks, &c.tasks[i])
+			}
+
+			s := ServiceStateLister{
+				Log:          hclog.NewNullLogger(),
+				ConsulClient: consulClient,
+				ECSClient: &mocks.ECSClient{
+					Tasks:           tasks,
+					PaginateResults: c.paginateResults,
+				}}
+
+			resources, err := s.List()
+
+			require.NoError(t, err)
+
+			serviceStates := make(map[string]ServiceState)
+
+			for _, resource := range resources {
+				serviceInfo := resource.(*ServiceInfo)
+
+				// only set the expected acl token fields
+				var tokens []*api.ACLTokenListEntry
+				for _, token := range serviceInfo.ServiceState.ACLTokens {
+					tokens = append(tokens, &api.ACLTokenListEntry{
+						Description:       token.Description,
+						ServiceIdentities: token.ServiceIdentities,
+					})
+				}
+				serviceInfo.ServiceState.ACLTokens = tokens
+				serviceStates[serviceInfo.ServiceName] = serviceInfo.ServiceState
+			}
+
+			require.Equal(t, c.expected, serviceStates)
+		})
+	}
+}
+
+func TestReconcile(t *testing.T) {
+	cluster := "test-cluster"
+	aclToken1 := &api.ACLToken{
+		Description:       fmt.Sprintf("Token for service1 service\n%s: %s", clusterTag, cluster),
+		ServiceIdentities: []*api.ACLServiceIdentity{{ServiceName: "service1"}},
+	}
+
+	cases := map[string]struct {
+		tasks          bool
+		aclTokens      []*api.ACLToken
+		sutServiceName string
+		expected       []*api.ACLToken
 	}{
-		"without pagination": {},
-		"with pagination":    {paginateResults: true},
+		"token should be deleted": {
+			tasks:          false,
+			aclTokens:      []*api.ACLToken{aclToken1},
+			sutServiceName: "service1",
+			expected:       nil,
+		},
+		"token should be added": {
+			tasks:          true,
+			aclTokens:      []*api.ACLToken{},
+			sutServiceName: "service1",
+			expected:       []*api.ACLToken{aclToken1},
+		},
+		"Does nothing when a task is running and a token exists for a given service": {
```

Comment:
should we also add a case that it does nothing when there are no tasks and tokens?